### PR TITLE
Curve AMO

### PR DESCRIPTION
**Pull request by clement-ux**
## Description 
Create a new AMO for Base, on Curve.
Almost same AMO as [current Convex AMO](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/strategies/ConvexEthMetaStrategy.sol) (on mainnet), with following exceptions:
- Depositing LPs on Curve gauge instead of Convex.
- Use WETH instead of raw ETH.
- Add `_solvencyAssert()` from AerodromeAMO [logic](https://github.com/OriginProtocol/origin-dollar/blob/1817476823da3efb7bc9b44020545f09567d11c9/contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol#L457-L475).

## Code Change Checklist

To be completed before internal review begins:

- [x]  The contract code is complete
- [x]  Executable deployment file
- [x]  Fork tests that test after the deployment file runs
- [x]  Unit tests *if needed
- [x]  The owner has done a [full checklist review](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) of the code + tests

Internal review:

- [ ] Two approvals by internal reviewers

### CODE REVIEW

**Comment by clement-ux** [comment]
## Author code review

## Requirements

_What is the PR trying to do? Is this the right thing? Are there bugs in the requirements?_

## Easy Checks

#### Authentication

- [ ] Never use tx.origin
- [ ] Every external/public function is supposed to be externally accessible
- [ ] Every external/public function has the correct authentication

#### Ethereum

- [ ] Contract does not send or receive Ethereum.
- [ ] Contract has no payable methods.
- [ ] Contract is not vulnerable to being sent self destruct ETH

#### Cryptographic code

- [ ] This contract code does not roll it's own crypto.
- [ ] No signature checks without reverting on a 0x00 result.
- [ ] No signed data could be used in a replay attack, on our contract or others.

#### Gas problems

- [ ] Contracts with for loops must have either:
    - [ ] A way to remove items
    - [ ] Can be upgraded to get unstuck
    - [ ] Size can only controlled by admins
- [ ] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### Black magic

- [ ] Does not contain `selfdestruct`
- [ ] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [ ] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [ ] Code is solidity version >= 0.8.0
- [ ] All for loops use uint256

#### Proxy

- [ ] No storage variable initialized at definition when contract used as a proxy implementation.

#### Events
- [ ] All state changing functions emit events

## Medium Checks

#### Rounding and casts
- [ ] Contract rounds in the protocols favor
- [ ] Contract does not have bugs from loosing rounding precision
- [ ] Code correctly multiplies before division
- [ ] Contract does not have bugs from zero or near zero amounts
- [ ] Safecast is aways used when casting

#### Dependencies
- [ ] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes.
- [ ] If OpenZeppelin ACL roles are use review & enumerate all of them.
- [ ] Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used.

#### External calls

- [ ] Contract addresses passed in are validated
- [ ] No unsafe external calls
- [ ] Reentrancy guards on all state changing functions
    - [ ] Still doesn't protect against external contracts changing the state of the world if they are called.
- [ ] No malicious behaviors
- [ ] Low level call() must require success.
- [ ] No slippage attacks (we need to validate expected tokens received)
- [ ] Oracles, one of:
  - [ ] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [ ] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting

#### Tests

  - [ ] Each publicly callable method has a test
  - [ ] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested
  - [ ] If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing.

#### Deploy

- [ ] Deployer permissions are removed after deploy

## Strategy Specific

_Remove this section if the code being reviewed is not a strategy._

#### Strategy checks

- [ ] Check balance cannot be manipulated up AND down by an attacker
- [ ] No read only reentrancy on downstream protocols during checkBalance
- [ ] All reward tokens are collected
- [ ] The harvester can sell all reward tokens
- [ ] No funds are left in the contract that should not be as a result of depositing or withdrawing
- [ ] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [ ] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation.
- [ ] WithdrawAll() cannot be MEV'd
- [ ] Strategist cannot steal funds

#### Downstream

- [ ] We have monitoring on all backend protocol's governances
- [ ] We have monitoring on a pauses in all downstream systems

## Thinking

#### Logic

  _Are there bugs in the logic?_

  - [ ] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).

#### Deployment Considerations

_Are there things that must be done on deploy, or in the wider ecosystem for this code to work. Are they done?_

#### Internal State

- What can be always said about relationships between stored state
- What must hold true about state before a function can run correctly (preconditions)
- What must hold true about the return or any changes to state after a function has run.

Does this code do that?

#### Attack

_What could the impacts of code failure in this code be._

_What conditions could cause this code to fail if they were not true._

_Does this code successfully block all attacks._

#### Flavor

_Could this code be simpler?_

_Could this code be less vulnerable to other code behaving weirdly?_

**Review by sparrowDom — CHANGES_REQUESTED** [review]
The strategy contract shares a lot of similarities with the `ConvexEthMetaStrategy` contract. I think we should refactor the common pieces of code into a separate class

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+        uint256[] memory balancesAfter = curvePool.get_balances();
+        // diff = ETH balance - OETH balance
+        int256 diffAfter = int256(balancesAfter[ethCoinIndex]) -
+            int256(balancesAfter[oethCoinIndex]);
+
+        if (diffBefore <= 0) {
+            // If the pool was originally imbalanced in favor of OETH, then
+            // we want to check that the pool is now more balanced
+            require(diffAfter <= 0, "OTokens overshot peg");
+            require(diffBefore < diffAfter, "OTokens balance worse");
```
[the remark this answers is not in the document]
Comment (sparrowDom) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:104]: This should probably be a: 
if (diffBefore == 0)
...
else if(diffBefore < 0)
...
else if(diffBefore > 0)
...

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+
+import { IERC20, InitializableAbstractStrategy } from "../utils/InitializableAbstractStrategy.sol";
+import { StableMath } from "../utils/StableMath.sol";
+import { IVault } from "../interfaces/IVault.sol";
+import { IWETH9 } from "../interfaces/IWETH9.sol";
+import { ICurveStableSwapNG } from "../interfaces/ICurveStableSwapNG.sol";
+import { ICurveXChainLiquidityGauge } from "../interfaces/ICurveXChainLiquidityGauge.sol";
+import { IChildLiquidityGaugeFactory } from "../interfaces/IChildLiquidityGaugeFactory.sol";
+
+contract BaseCurveAMOStrategy is InitializableAbstractStrategy {
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:19]: All curve pools are different, so it is probably okay to be specific.

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+    using StableMath for uint256;
+
+    /**
+     * @dev a threshold under which the contract no longer allows for the protocol to manually rebalance.
+     *      Guarding against a strategist / guardian being taken over and with multiple transactions
+     *      draining the protocol funds.
+     */
+    uint256 public constant SOLVENCY_THRESHOLD = 0.998 ether;
+
+    uint256 public constant MAX_SLIPPAGE = 1e16; // 1%, same as the Curve UI
```
[the remark this answers is not in the document]
Comment (sparrowDom) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:29]: I guess we have the vaultValueChecker that guards against the slippage issues. But that check is manually enforced and can easily be overlooked by reviewers and the person making the proposal

[the remark this answers is not in the document]
Comment (clement-ux) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:29]: Changed in this commit: https://github.com/OriginProtocol/origin-dollar/pull/2384/commits/6df069e25fab794d34f207b71195e9cda1016844.
MAX_SLIPPAGE is now dynamic.

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+            msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Caller is not the Strategist"
+        );
+        _;
+    }
+
+    /**
+     * @dev Checks the Curve pool's balances have improved and the balances
+     * have not tipped to the other side.
+     * This modifier only works on functions that do a single sided add or remove.
```
[the remark this answers is not in the document]
Comment (clement-ux) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:80]: Fixed in this commit: https://github.com/OriginProtocol/origin-dollar/pull/2384/commits/61cd6c982641cb00f21b8b3ed4bf039518198ee2.

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (sparrowDom) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:29]: nice thanks

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (naddison36) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:104]: Adding the explicit `if (diffBefore == 0)` does make it clearer. Thanks

**Review by naddison36 — DISMISSED** [review]

File: contracts/utils/addresses.js [hunk 5]
```diff
@@ -350,6 +350,14 @@ addresses.base.oethbBribesContract =
 
 addresses.base.OZRelayerAddress = "0xc0D6fa24D135c006dE5B8b2955935466A03D920a";
 
+// Base Curve
+addresses.base.CRV = "0x8Ee73c484A26e0A5df2Ee2a4960B789967dd0415";
+addresses.base.OETHb_WETH = {};
```
[the remark this answers is not in the document]
Comment (sparrowDom) [reply contracts/utils/addresses.js:355]: I like this one as well, using it in the future

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+    ICurveXChainLiquidityGauge public immutable gauge;
+
+    /**
+     * @notice Address of the Child Liquidity Gauge Factory contract.
+     */
+    IChildLiquidityGaugeFactory public immutable gaugeFactory;
+
+    // Ordered list of pool assets
+    uint128 public constant oethCoinIndex = 1;
+    uint128 public constant ethCoinIndex = 0;
```
Comment (sparrowDom) [inline contracts/contracts/strategies/BaseCurveAMOStrategy.sol:64]: confirming the order is correct:
Screenshot 2025-02-07 at 09 56 16
Screenshot 2025-02-07 at 09 56 05

File: contracts/contracts/strategies/BaseCurveAMOStrategy.sol [hunk 7]
```diff
@@ -0,0 +1,641 @@
...
```
[the remark this answers is not in the document]
Comment (sparrowDom) [reply contracts/contracts/strategies/BaseCurveAMOStrategy.sol:64]: I don't remember why suggested the rename to `ethCoinIndex`. The token in the pool is obviously WETH and should probably named `wethCoinIndex`. Sorry about that

**Review by DanielVF — DISMISSED** [review]

**Comment by sparrowDom** [comment]
## Requirements

PR is creating an AMO strategy on Base using the OToken & WETH assets.

## Easy Checks
#### Authentication

- [x] Never use tx.origin
- [] Every external/public function is supposed to be externally accessible
  🟡  checkBalance should be `external` instead of `public`
- [x] Every external/public function has the correct authentication

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [X] Contract has no payable methods.
- [x] Contract is not vulnerable to being sent self destruct ETH

#### Cryptographic code
no own crypto 

#### Gas problems
no loops

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [X] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.
- [ ] 🟡 set the 0 governor to implementation contract (by calling `_setGovernor(address(0));` in the constructor)

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding and casts
- [x] Contract rounds in the protocols favor 
- [x] Contract does not have bugs from loosing rounding precision
- [x] Code correctly multiplies before division
- [x] Contract does not have bugs from zero or near zero amounts
- [x] Safecast is aways used when casting

#### Dependencies
- [x] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes.
- [ ] ~~If OpenZeppelin ACL roles are use review & enumerate all of them.~~
- [ ] ~~Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used.~~

#### External calls

- [x] Contract addresses passed in are validated
- [x] No unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [x] Still doesn't protect against external contracts changing the state of the world if they are called.
- [x] No malicious behaviors
- [x] Low level call() must require success.
- [ ] No slippage attacks (we need to validate expected tokens received)
  - 🟠 In the first Curve implementation we checked the pool code to confirm that the `minAmount` passed to `add_liquidity` would be respected. If we intend to use the same code for multiple AMOs we should manually check after `add_liquidity` call that expected amount of LP tokens has been received: 
  ```
  require(lpDeposited >= minMintAmount, "Min LP amount error");
  ``` 
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting

#### Tests

  - [ ] Each publicly callable method has a test
  - [ ] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested
  - [ ] If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing.

#### Deploy

- [x] Deployer permissions are removed after deploy

## Strategy Specific

#### Strategy checks

- [x] Check balance cannot be manipulated up AND down by an attacker
- [x] No read only reentrancy on downstream protocols during checkBalance
- [x] All reward tokens are collected
- [x] The harvester can sell all reward tokens
- [x] No funds are left in the contract that should not be as a result of depositing or withdrawing
- [x] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [x] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation.
- [x] WithdrawAll() cannot be MEV'd
- [x] Strategist cannot steal funds

#### Downstream

- [ ] We have monitoring on all backend protocol's governances
- [ ] We have monitoring on a pauses in all downstream systems

## Thinking

#### Logic

  _Are there bugs in the logic?_

  - [x] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).

#### Deployment Considerations

nothing special needs to happen on deploy

#### Internal State

The code makes sure that the AMO position (and pool's self token ratio) in the pool after depositing always has at least 1:1 (OETH : WETH) ratio and in the worst case 2:1 (OETH : WETH) ratio.

Does this code do that?
Yes 

#### Attack

If there were errors in slippage / expected token configurations we could be loosing user funds.

A faulty `get_virtual_price` on the curve pool could be problematic. 

No

#### Flavor

Code is written quite nicely. We've tried to refactor bits of it, but it turned out to be worse. There might be room for improvement, to put some of the more math intensive bits of code into a library.

**Review by sparrowDom — DISMISSED** [review]
